opengovsg/FormSG
# feat: Add submission limits for storage mode form submissions

## Problem
This pull request adds a submission limits feature for storage.

Closes #148

## Solution
Instead of using the `formStatisticsTotal` collection, I am simply querying `SubmissionModel.countDocuments({ form: form._id })` for the number of submissions up till now. I think this is probably fine as the submission model has an index on the `form` field and we are already using that to count the total number of submissions for export and attachment downloads and those seem reasonably fast.

Otherwise, we are allowing users to specify a submission limit, but will simply show them an error message instead of attempting to deactivate the form, which is more complicated and requires us to touch the form document itself instead of simply checking for number of submissions before we render the form and before the form is submitted.

This also doesn't require re-CAPTCHA to be turned on (maybe we should?).

**Features**:

- New field to let the user specify how many responses they want.
- Checks for submission limits and shows the user an error message if they visit a form that has exceeded the maximum limit

## Before & After Screenshots

![image](https://user-images.githubusercontent.com/691628/107134484-57fda580-68a7-11eb-9947-d69d7536542e.png)
![image](https://user-images.githubusercontent.com/691628/107134499-7a8fbe80-68a7-11eb-9e29-51672a8c8d2a.png)

## Tests

- Create a form, and set the limit to zero. 
- Verify that the form is not visible. 
- Set the limit to 1 and submit the form.
- Verify that the submission goes through
- Visit the form again and make sure that the form is not visible any more.
- Set the limit to 2 and open two tabs.
- Attempt to submit the form in both tabs. Only one should work.

## Deploy Notes
This introduces a new `submissionLimit` field in the `Form` collection.
